### PR TITLE
Pipeline response object

### PR DESCRIPTION
**Pull request by steven-foster**
This allows the pipeline step to return an object with some response information. The main reason for this addition is to allow subsequent messages to be added as a thread on an initial message. It builds upon the thread support added in a recent PR.

Example usage in a Jenkinsfile:
`response = slackSend botUser: true, channel: 'general', message: 'Hello'
slackSend botUser: true, channel: "${response.threadId}", message: 'this is a thread response of the initial message'`

### CODE REVIEW

File: src/main/java/jenkins/plugins/slack/workflow/SlackResponse.java [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+import org.jenkinsci.plugins.scriptsecurity.sandbox.whitelists.Whitelisted;
+import org.json.JSONObject;
+
+import java.io.Serializable;
+
+public class SlackResponse implements Serializable {
+    private String channelId;
+    private String ts;
+
+    public SlackResponse() {
```
Comment (timja) [inline src/main/java/jenkins/plugins/slack/workflow/SlackResponse.java]: why is this here?

File: src/test/java/jenkins/plugins/slack/workflow/SlackSendStepTest.java [hunk 2]
```diff
@@ -51,6 +53,14 @@
     @Mock
     SlackNotifier.DescriptorImpl slackDescMock;
 
+    String savedResponse = "{\"ok\":true,\"channel\":\"F4KE1DABC\",\"ts\":\"1543931401.000500\"," +
```
Comment (timja) [inline src/test/java/jenkins/plugins/slack/workflow/SlackSendStepTest.java]: nit, to clean it up, could you put it in a json file and load it as a resource?
no need for string escaping then and would be much easier to read?

File: src/test/java/jenkins/plugins/slack/workflow/SlackSendStepTest.java [hunk 3]
```diff
@@ -204,4 +207,40 @@ public void testNonNullEmptyColor() throws Exception {
...
+        );
+        when(slackServiceMock.getResponseString()).thenReturn(savedResponse);
+        when(slackServiceMock.publish(anyString(), anyString())).thenReturn(true);
+
+        SlackResponse response = stepExecution.run();
+        String expectedId = "F4KE1DABC";
+        String expectedTs = "1543931401.000500";
+        assertNotNull(response);
+        assertEquals(expectedId, response.getChannelId());
+        assertEquals(expectedTs, response.getTs());
```
Comment (timja) [inline src/test/java/jenkins/plugins/slack/workflow/SlackSendStepTest.java]: One more assertion for the thread id on the slack response?

**Comment by timja** [comment]
Also build is failing

File: src/main/java/jenkins/plugins/slack/workflow/SlackResponse.java [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+import org.json.JSONObject;
+
+import java.io.Serializable;
+
+public class SlackResponse implements Serializable {
+    private String channelId;
+    private String ts;
+
+    public SlackResponse(String slackResponseString) {
+        if (StringUtils.isEmpty(slackResponseString)) {
```
Comment (timja) [inline src/main/java/jenkins/plugins/slack/workflow/SlackResponse.java]: I don't think this change is quite what you wanted 😉 
1. "ok" is the text returned for non bot users, but yes its correctly null as your new test demonstrates
2. now the case that you want doesn't work as this code is only invoked when the response is empty

maybe 
```suggestion
        if (!"ok".equals(slackResponseString)) { // non botUsers respond with only 'ok'
```

Comment (steven-foster) [reply src/main/java/jenkins/plugins/slack/workflow/SlackResponse.java]: Whoops, I seem to have mixed up two different approaches somehow... thanks for your patience.

I was trying to make the response string only populate when bot user is turned on, and in other cases return a SlackResponse object with null fields. I would be worried about any other unknown cases if I simply checked for "ok".

File: src/test/java/jenkins/plugins/slack/workflow/SlackSendStepTest.java [hunk 5]
```diff
@@ -209,4 +212,77 @@ public void testNonNullEmptyColor() throws Exception {
...
+
+        when(Jenkins.getActiveInstance()).thenReturn(jenkins);
+
+        stepExecution.listener = taskListenerMock;
+        when(taskListenerMock.getLogger()).thenReturn(printStreamMock);
+        doNothing().when(printStreamMock).println();
+
+        when(stepExecution.getSlackService(anyString(), anyString(), anyString(), anyString(), anyBoolean(), anyString())).thenReturn(slackServiceMock);
+
+        when(slackServiceMock.getResponseString()).thenReturn(null);
```
Comment (timja) [inline src/test/java/jenkins/plugins/slack/workflow/SlackSendStepTest.java]: doesn't this return `ok` though?

File: src/main/java/jenkins/plugins/slack/workflow/SlackResponse.java [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+import org.json.JSONObject;
+
+import java.io.Serializable;
+
+public class SlackResponse implements Serializable {
+    private String channelId;
+    private String ts;
+
+    public SlackResponse(String slackResponseString) {
+        if (!StringUtils.isEmpty(slackResponseString)) {
```
Comment (timja) [inline src/main/java/jenkins/plugins/slack/workflow/SlackResponse.java]: won't this be `ok` for a non bot user?

Comment (steven-foster) [reply src/main/java/jenkins/plugins/slack/workflow/SlackResponse.java]: currently it only registers the response for a bot user, so this should be `null` for a non bot user. This is a precaution against any other case that I might not be aware of. I can change it to check for `ok` explicitly, I just wasn't sure if that would cover everything

**Comment by timja** [comment]
How have you setup your bot user?
I don't use this feature myself...

I generated an api token here:
https://api.slack.com/custom-integrations/legacy-tokens

Set it up as a secret text credential

Configured in the UI with botUser true and the new credential

Tested it in a public channel, test says success, but no test message arrives
Tested by dm'ing me, test message arrives fine

Tried to run it in a pipeline and there was a null response

**Comment by steven-foster** [comment]
yeah it took some figuring out, it seems like bot users are the most recent and recommended way to do things though.

 The normal Jenkins slack app only seems to handle non-bot user messages. You have to create an app in your slack workspace to own the bot from https://api.slack.com/apps. The create new app wizard should offer bot user as a suggestion for your app.

You create a bot with whatever name you need, then go to OAuth & Permissions / Scopes and add the chat:write:bot/Send messages as... permission. Then install the app to the workspace, which will give you a Bot User OAuth Access Token. You might also have to actually invite the bot to the channel in slack.

**Comment by timja** [comment]
Thanks for the instructions, I probably won't have time today to set that up, but will get to it tomorrow.
dw it'll be in soon
